pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -53,12 +58,26 @@ func (f *Filter) GetUnderutilizedNodes(ctx context.Context) ([]*v1.Node, error)
 func (f *Filter) getPodsOnNode(ctx context.Context, nodeName string) (*v1.PodList, error) {
 	pods := &v1.PodList{}
 	if err := f.kubeClient.List(ctx, pods, client.MatchingFields{"spec.nodeName": nodeName}); err != nil {
-		return nil, fmt.Errorf("listing unscheduled pods, %w", err)
+		return nil, fmt.Errorf("listing pods on nodes, %w", err)
 	}
 	return pods, nil
 }
 
-// TODO: implement underutilized function
-func (f *Filter) isUnderutilized(*v1.Node, *v1.PodList) bool {
-	return false
+// TODO: implement underutilized function (some generalized predicate)
+func (f *Filter) isUnderutilized(pods *v1.PodList) bool {
+	counter := 0
+	for _, pod := range pods.Items {
+		if podUtil.IsNotIgnored(&pod) {
+			counter += 1
+		}
+	}
+	return counter == 0
+}
+
+func (f *Filter) getNodesWithLabels(ctx context.Context, labelKeys []string) (*v1.NodeList, error) {
```

Comment:
This is a nice low level function, but we should wrap it with more useful ones.

```
func getNodes(ctx context.Context, provisioner v1alpha1.Provisioner)
func getTTLedNodes(ctx context.Context, provisioner v1alpha1.Provisioner)
func getCordonedNodes(ctx context.Context, provisioner v1alpha1.Provisioner)
func getDrainingNodes(ctx context.Context, provisioner v1alpha1.Provisioner)
```

Reply:
Good idea. Will do